### ISSUE TITLE
PublishStorageDeals Optimizations

### ISSUE THREAD

**Issue by ZenGround0** [filecoin-project/specs-actors#1305]
Anecdotally the PublishStorageDeals method is very expensive in terms of gas cost.  This is becoming especially apparent in light of more accurate gas assessments.

Some of this is unavoidable since storing deal metadata on chain is expensive.  However it is likely that logic specific optimizations will help bring down costs.  This issue is to profile and explore these optimizations.  This issue does not cover HAMT / AMT parameter tuning and performance improvements which will also have a positive impact on this work.

**Comment by ZenGround0** [comment]
Notes from conversation with @Stebalien 

Gas trace used in investigation is on ipfs at: `QmTqQES1LANjysAfKsNrAQUT3Qmb2ofB1Dgc7rpeZXXMFH`

PublishStorageDeals (PSD) gas cost is dominated by many series of Ipld Puts without a single culprit. The biggest single gas charge is signature verification but this is less that 1/100 of gas charge on typical PSD messages.

-- Synergy with existing datastructures work -- 
We expect to see some small improvements from fixing the HAMT to avoid duplicate writes but probably not on the order of 40-50% max possible savings due to the access patterns. Need to measure this once we have improved HAMT.  Also removing duplicate writes on AMT/HAMT is needed to unlock the benefits of using existing PSD batching because otherwise we will incur duplicate writes for individual deals.

-- New Solutions -- 
1. PendingProposals stores the proposal data (empirically ~153 byte values) but it is not needed.  We should just use cid keys.
2. DealOpsByEpoch
Good ideas:
  a. The HAMT HAMT structure is not ideal, it should be a HAMT of AMTs because we can pack items sequentially: key is array index, value is deal ID.  While cbor array would save space favor AMT over cbor array so that adding new ids doesn't require loading the whole set of deal ids.
  b. We should tune the inner AMT's width based on size of objects
  c. Add deals to DealOpsByEpoch in batches so that we load fewer of the inner AMTs.  This will increase variance in distribution of number of deals processed per epoch by X with X being the batch size which seems ok.  Complication: all deals in batch must be within one day away from the DOBE key they are placed at which constrains which deals can be put in a batch.

Less good / Discarded ideas:
  d. Instead of using a giant HAMT with epoch keys we could use something closer to miner deadline where the key is epoch % 2880.  The problem here is that deals can start arbitrarily far into the future which means we will have a mix of deals that should be processed and deals that should not be processed in the list.  There are maybe ways to deal with this but @Stebalien is pessimistic they will be sufficient
  e. Instead of tracking which deals to process in DOBE just select a random subset of live deal ids to process every cron tick
  f. Instead of a raw AMT as the inner object of DOBE use an AMT with cbor array leavs to shard a cbor array.  Variance of gas cost increases but average cost could go down
  g. Intrusive linked list (I just got the name, didn't grok the concept, but apparently its not worth doing anyway)
  h. use a new ipld datastructure for ordered map to merge together sets in set multimap to the toplevel.  Turns out this is not useful since we are not doing lookups on the sets, we are iterating them as a list.
  
d, e, and f still seem worth thinking about a bit from my perspective.  f, g, h just included for completeness / in case they give others ideas.

**Comment by Stebalien** [comment]
> PendingProposals stores the proposal data (empirically ~153 byte values) but it is not needed. We should just use cid keys.

Note: this is actually worse than that because HAMTs can have up to 3*32 elements at a level. A 153 byte write with a 36 byte key (cid) can turn into `(153+36)*3*32 = 17KiB` write (very approximately, not sure if the empirical value size estimate included the key). Getting rid of the value will reduce this to 3.5KiB which isn't great but isn't terrible either.

This _doesn't_ apply to the actual deals AMT because AMTs only have 8 entries and don't explicitly store keys.

> Instead of using a giant HAMT with epoch keys we could use something closer to miner deadline where the key is epoch % 2880. The problem here is that deals can start arbitrarily far into the future which means we will have a mix of deals that should be processed and deals that should not be processed in the list. There are maybe ways to deal with this but @Stebalien is pessimistic they will be sufficient
e. Instead of tracking which deals to process in DOBE just select a random subset of live deal ids to process every cron tick

If we do set a maximum start date, this may actually be doable. But it may not be worth it.

>  g. Intrusive linked list (I just got the name, didn't grok the concept, but apparently its not worth doing anyway)

Basically, given that we just need to iterate through deals at every epoch and we don't need to do lookups, we could link up the deals into an intrusive linked list (linking each deal to the next by deal ID). But this really seems like something that could go terribly wrong.

**Comment by ZenGround0** [comment]
> very approximately, not sure if the empirical value size estimate included the key

Woops no, its actually 151 bytes for value and 38 bytes for key

**Comment by acruikshank** [comment]
I modified the `SetMultiset` data structure used by `DealOpsByEpoch` then ran simulations to measure the store access required by PublishStorageDeals. The results aren't looking great.

Before the change (HAMT of HAMTs):
```fil/2/storagemarket:4: calls: 61  gets: 1300  puts: 1008  read: 1159088  written: 1167608  avg gets: 21.31, avg puts: 16.52```

After the change (HAMT of AMTs):
```fil/2/storagemarket:4: calls: 61  gets: 1308  puts: 1194  read: 1167558  written: 1176217  avg gets: 21.44, avg puts: 19.57```

This suggests that switching to an AMT __significantly increases__ both write count and total write size. It's possible tuning the AMT bitwidth might make this better, but the HAMT version might also be tuned for better performance. I recommend we pursue other optimizations.

**Comment by acruikshank** [comment]
I actually implemented it incorrectly the first time (using the AMT as a sparse array indexed by Deal IDs, rather than a simple list of deal IDs). Implementing it correctly improves the results from: 
```fil/2/storagemarket:4: calls: 61 gets: 1300 puts: 1008 read: 1159088 written: 1167608 avg gets: 21.31, avg puts: 16.52```

To:
```fil/2/storagemarket:4: calls: 61  gets: 1300  puts: 1008  read: 1158860  written: 1167377  avg gets: 21.31, avg puts: 16.52```

This means the HAMT and AMT versions have cause the same number of reads and writes and total read and write size is reduced, but only by 0.02%. So it's still debatable whether this is worth doing.

**Comment by acruikshank** [comment]
Since I was already in this, I decided just to see what happens if we used a simple CBOR array instead of a HAMT or AMT in the SetMultiset. The stats look like this:

```fil/2/storagemarket:4: calls: 61  gets: 1297  puts: 946  read: 1074600  written: 1083064  avg gets: 21.26, avg puts: 15.51```

That's a 6% reduction in writes (which we expect by getting rid of all the (H)AMT nodes, but also reduces total write size (for the whole PublishStorageDeal call) by 7%. This is surprising because we would expect having to rewrite the entire array for every append would be wasteful, but for the deal ops in the simulation, it appears that it beats the overhead of using a HAMT or AMT. This is probably because we are looking at storing a list of 10-200 integers.

I say it's worth more investigation to determine in what domains (H)AMT performance exceeds CBOR arrays for task like this where we only need a list of items.

**Comment by Stebalien** [comment]
I guess the question is: are these simulations accurate in the long run? Assuming a million deals, I'd expect each of these arrays to be ~350 integers long.

Also, what happens if we increase the AMT bitwdith? Unfortunately, we _really_ want the AMT bitwidth for intermediate nodes to be small, but large for leafs. The only way to do that right now is to store arrays of numbers in the leaves.

**Comment by anorth** [comment]
We should probably be aiming at 10s of millions of deals at least.

Note that a major use of this structure is in cron, where an entire queue entry will be popped off the outer queue, traversed, and appended under a different epoch key. No-one pays the gas for this, but we mustn't optimize too much for the random insertion that we will cause big problems in cron. Note that we have room to increase the cardinality of the outer queue (100 -> 1000?) if that helps. The effect of this is to decrease granularity of deal payments and some termination edge cases.

**Comment by anorth** [comment]
We believe #1316 hit the best of low-hanging fruit here. That's not to say more optimization isn't possible, but it will likely involve a larger reworking of state. Closing this for now.
